src/rpc/blockchain.cpp

```diff
@@ -1616,6 +1616,9 @@ static RPCHelpMan reconsiderblock()
         throw JSONRPCError(RPC_DATABASE_ERROR, state.ToString());
     }
 
+    // Update best header
+    WITH_LOCK(::cs_main, chainman.m_best_header = chainman.ActiveChain().Tip());
```

Comment:
> Conceptually, this seems the wrong place for the fix. It would be nice if this was handled internally in chainman correctly.

I did that initially (see #26260 - https://github.com/bitcoin/bitcoin/pull/26260#issuecomment-1270121139) but.. conceptually, all what is inside `reconsiderblock` seems to be at the wrong place for me too. We are resetting a block index flag at the RPC level. And, ideally, we should encapsulate all this code within chainman too.

Also, see https://github.com/bitcoin/bitcoin/pull/26260#pullrequestreview-1132786161 for the places where we set this field. We either set it when a new block is received, during startup or manually inside `invalidateblock()`. Thus why I think that setting it inside `reconsiderblock` is also "fine" (it is what we are already doing for its counterpart anyway).

> Also, given that cs_main is held only for this line, a race exits, where the getchainstates RPC is called before this line, and still returns the wrong result.

True. But, as RPC calls are blocking and this RPC command is usually a one-time only not critical call, I think that it is a harmful race.